openedx/core/djangoapps/credentials/tasks/v1/tasks.py

```diff
@@ -54,3 +75,288 @@ def send_grade_to_credentials(self, username, course_run_key, verified, letter_g
+            continue
+
+        logger.info(
+            "Handling grade changes %d for grade %s",
+            i, gradestr(grade),
+        )
+
+        user = User.objects.get(id=grade.user_id)
+
+        # Grab mode/status from cert call
+        key = (user.id, str(grade.course_id))
+        cert_info = course_cert_info.get(key, {})
+        mode = cert_info.get('mode', None)
+        status = cert_info.get('status', None)
+
+        send_grade_if_interesting(
+            user,
+            grade.course_id,
+            mode,
+            status,
+            grade.letter_grade,
+            grade.percent_grade,
+            verbose=verbose
+        )
+
+
+def paged_query(queryset, delay, page_size):
+    """
+    A generator that iterates through a queryset but only resolves chunks of it at once, to avoid overwhelming memory
+    with a giant query. Also adds an optional delay between yields, to help with load.
+    """
+    count = queryset.count()
+    pages = int(math.ceil(count / page_size))
+
+    for page in range(pages):
+        page_start = page * page_size
+        page_end = page_start + page_size
+        subquery = queryset[page_start:page_end]
+
+        if delay and page:
+            time.sleep(delay)
+        index = 0
+        try:
+            for item in subquery.iterator():
+                index += 1
+                yield page_start + index, item
+        except OperationalError:
+            # When running the notify_credentials command locally there is an
+            # OperationalError thrown by MySQL when there are no more results
+            # available for the queryset iterator. This change catches that exception,
+            # checks state, and then logs according to that state. This code runs in
+            # production without issue. This changes allows for the code to be run
+            # locally without a separate code path.
+            if index == count:
+                logger.info('OperationalError Exception caught, all known results processed in paged_query')
+            else:
+                logger.warning('OperationalError Exception caught, it is possible some results were missed')
+            continue
+
+
+def log_dry_run(certs, grades):
+    """Give a preview of what certs/grades we will handle."""
+    logger.info("DRY-RUN: This task would have handled changes for...")
+    ITEMS_TO_SHOW = 10
+
+    logger.info(f"{certs.count()} Certificates:")
+    for cert in certs[:ITEMS_TO_SHOW]:
+        logger.info(f"   {certstr(cert)}")
+    if certs.count() > ITEMS_TO_SHOW:
+        logger.info(f"    (+ {certs.count() - ITEMS_TO_SHOW} more)")
+
+    logger.info(f"{grades.count()} Grades:")
+    for grade in grades[:ITEMS_TO_SHOW]:
+        logger.info(f"   {gradestr(grade)}")
+    if grades.count() > ITEMS_TO_SHOW:
+        logger.info(f"    (+ {grades.count() - ITEMS_TO_SHOW} more)")
+
+
+def certstr(cert):
+    return f'{cert.course_id} for user {cert.user.id}'
```

Comment:
This changed: this used to log `cert.user.username`, but I changed it to `id` so we’re not putting usernames in our logs for PII concerns. (The equivalent `gradestr` below just logs the id.)

Reply:
Nit: Should we also update the "Example Usage" comments in `notify_credentials.py` to reflect that this will now log user id over username?
```
        3 Certificates:
            course-v1:edX+RecordsSelfPaced+1 for user records_one_cert
            course-v1:edX+RecordsSelfPaced+1 for user records
            course-v1:edX+RecordsSelfPaced+1 for user records_unverified
```
https://github.com/edx/edx-platform/blob/4462ed37b243f56c481e80c21988ecc7337a9786/openedx/core/djangoapps/credentials/management/commands/notify_credentials.py#L106-L109